### PR TITLE
[Proposal] Moves Padding Property to Container

### PR DESCRIPTION
**Pull request by RemieRichards**
* Moves ```Padding``` Property to Container 
* Removes ```PaddingContainer```
* Renames ```FlowContainer.Padding``` to ```FlowContainer.ElementPadding```

I'm proposing this due to problems/issues I had/discovered while creating a ```CheckBox``` (this is a bit of a ramble, I apologise):

A ```CheckBox``` is a simple ```AutoSizeContainer```, with a private ```FlowContainer``` for "labels", labels allow you to attach ```SpriteText``` or other ```Drawable```s to the ```CheckBox``` such that clicking on them also counts as a click on the ```CheckBox``` (a common functionality of checkboxes), However when I added ```PaddingContainer```s around these labels, so that they weren't glued to the CheckBox, this was problematic, ```PaddingContainer```s do not auto size, and as such remained at 0x0, preventing the ```FlowContainer``` and the ```CheckBox``` from resizing themselves based on the size of the labels, effectively masking them out and preventing the label-click feature, so instead of creating some Frankenstein class of ```AutoSizePaddingContainer```, I moved the ```Padding``` property+functionality down to ```Container```.

With ```Padding``` moved to ```Container```, ```PaddingContainer``` no longer needs to exist and as such, has been removed.

tl;dr:
```
new PaddingContainer
{
    Padding = new Padding
    {
        Left = 20,
    },
    Children = new Drawable[]
    {
        new SpriteText
        {
            Text = @"Some Padded Text",
        }
    }
}

//is now:

new SpriteText
{
    Padding = new Padding
    {
        Left = 20,
    },
    Text = @"Some Padded Text",
}
```

All tests, such as ```TestCasePadding``` work as expected still.

### CODE REVIEW

File: osu.Framework/Graphics/Containers/AutoSizeContainer.cs [hunk 1]
```diff
@@ -57,7 +57,7 @@ protected override Quad DrawQuadForBounds
                 if ((RelativeSizeAxes & Axes.Y) > 0)
                     maxBoundSize.Y = Size.Y;
 
-                return new Quad(0, 0, maxBoundSize.X, maxBoundSize.Y);
+				return new Quad(0, 0, maxBoundSize.X+Padding.Left-Padding.Right, maxBoundSize.Y+Padding.Top-Padding.Bottom);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/AutoSizeContainer.cs]: incorrect tabs here too

Comment (peppy) [inline osu.Framework/Graphics/Containers/AutoSizeContainer.cs]: i'm pretty sure this should be + Padding.Left + Padding.Right?
also there should be spaces on either side of the `+`

File: osu.Framework/Graphics/Containers/Container.cs [hunk 2]
```diff
@@ -88,12 +119,40 @@ public Container()
         /// <summary>
         /// Scale which is only applied to Children.
         /// </summary>
-        internal virtual Vector2 ChildOffset => Vector2.Zero;
+        internal virtual Vector2 ChildOffset => new Vector2(Padding.Left + Margin.Left, Padding.Top + Margin.Top);
+
+        /// <summary>
+        /// Because of our custom DrawQuad implementation below, we want to expose the *base* DrawQuad when something requests our bounds.
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/Container.cs]: I don't think <summary> tags are proper for explanatory comments. :P

File: osu.Framework/Graphics/Containers/Container.cs [hunk 3]
```diff
@@ -88,12 +119,40 @@ public Container()
...
+                    return new Quad(s.X, 0, -s.X, s.Y);
+                return new Quad(0, s.Y, s.X, -s.Y);
+            }
+        }
 
         /// <summary>
         /// The Size (coordinate space) revealed to Children.
         /// </summary>
-        internal virtual Vector2 ChildSize => Size;
+        internal virtual Vector2 ChildSize => base.Size - new Vector2(Padding.TotalHorizontal + Margin.TotalHorizontal, Padding.TotalVertical + Margin.TotalVertical);
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/Container.cs]: I'm not sure I understand why Margin has to be subtracted when you already use `base.Size` which doesn't have margin applied in the first place.

File: osu.Framework/Graphics/Containers/Container.cs [hunk 4]
```diff
@@ -80,6 +81,36 @@ public Container()
...
+        public Padding Padding
+        {
+            get { return padding; }
+            set
+            {
+                if (padding.Equals(value)) return;
+
+                padding = value;
+
+                Invalidate(Invalidation.Position);
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/Container.cs]: This affects children size, so children's "SizeInParentSpace" should also get invalidated. (Will not automatically be done by Invalidate(Invalidation.Position)). In fact, do we even need to invalidate our own position when Padding changes? I don't think so.

File: osu.Framework/Graphics/Containers/Container.cs [hunk 5]
```diff
@@ -373,4 +429,33 @@ public override Drawable DelayReset()
...
+        public float TotalHorizontal => Left + Right;
+
+        public float TotalVertical => Top + Bottom;
+
+        public Padding(float allSides)
+        {
+            Top = Left = Bottom = Right = allSides;
+        }
+
+        public int CompareTo(object obj)
```
Comment (huoyaoyuan) [inline osu.Framework/Graphics/Containers/Container.cs]: This method could probably be removed. `CompareTo` inherited from `ValueType` provides a correct behavior, and it should be never called manually because of the `IEquatable` implementation.

File: osu.Framework/Graphics/Containers/Container.cs [hunk 6]
```diff
@@ -373,4 +429,33 @@ public override Drawable DelayReset()
...
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool Equals(Padding other)
+        {
+            return Top == other.Top && Left == other.Left && Bottom == other.Bottom && Right == other.Right;
```
Comment (huoyaoyuan) [inline osu.Framework/Graphics/Containers/Container.cs]: Can also implement an `operator==` here to allow `==` grammar.

File: osu.Framework/Graphics/Containers/Container.cs (hunk 5 above)
Comment (peppy) [inline osu.Framework/Graphics/Containers/Container.cs]: Actually if you implement operator== you will get warnings about not implementing CompareTo and Equals(obj) (give it a try and you'll see).

Comment (huoyaoyuan) [inline osu.Framework/Graphics/Containers/Container.cs]: We've chosen `IEquatable` right? Remove `CompareTo` now won't get a warning.

Comment (peppy) [inline osu.Framework/Graphics/Containers/Container.cs]: yeah, that's the reason I used IEquatable<T>. not sure why CompareTo was added.

**Comment by SirCmpwn** [comment]
>Renames FlowContainer.Padding to FlowContainer.ElementPadding

Could drop this entirely and just have children set their margins. This is how it's done with flexbox in HTML, for instance.

**Comment by RemieRichards** [comment]
@SirCmpwn well, the name is going to be ```Spacing``` actually, that's the name Peppy wanted and is in the FlowContainer fix PR.

But yeah it could be dropped.
